controllers/pipelines/provider_controller.go

```diff
@@ -0,0 +1,220 @@
+				logger.Error(err, "unable to update provider service deployment", "deployment", desiredDeployment)
+				return ctrl.Result{}, err
+			}
+		}
+	} else {
+		if err = r.EC.Client.Create(ctx, desiredDeployment); err != nil {
+			logger.Error(err, "unable to create provider service deployment")
+			return ctrl.Result{}, err
+		}
+
+		logger.Info("created provider deployment", "deployment", desiredDeployment)
+	}
+
+	duration := time.Since(startTime)
+	logger.Info("reconciliation ended", logkeys.Duration, duration)
+
+	return ctrl.Result{}, nil
+}
+
+func (r *ProviderReconciler) SetupWithManager(mgr ctrl.Manager) error {
+	provider := &pipelinesv1.Provider{}
+	return ctrl.NewControllerManagedBy(mgr).
+		For(provider).
+		Owns(&appsv1.Deployment{}).
+		Complete(r)
+}
+
+func constructDeployment(provider *pipelinesv1.Provider, config config.KfpControllerConfigSpec) (*appsv1.Deployment, error) {
+	matchLabels := map[string]string{AppLabel: fmt.Sprintf("provider-%s", provider.Name)}
+	ownerLabels := map[string]string{OwnerNameLabel: fmt.Sprintf("provider-%s", provider.Name)}
+	deploymentLabels := pipelines.MapConcat(pipelines.MapConcat(config.DefaultProviderValues.Labels, matchLabels), ownerLabels)
+	replicas := int32(config.DefaultProviderValues.Replicas)
+
+	podTemplate := config.DefaultProviderValues.PodTemplateSpec
+	serviceContainerName := config.DefaultProviderValues.ServiceContainerName
+	podTemplate = populateServiceContainer(serviceContainerName, *podTemplate.DeepCopy(), provider)
+	podTemplate.Spec.ServiceAccountName = provider.Spec.ServiceAccount
+	podTemplate.ObjectMeta.Labels = pipelines.MapConcat(podTemplate.ObjectMeta.Labels, matchLabels)
+
+	deployment := &appsv1.Deployment{
+		TypeMeta: metav1.TypeMeta{
+			Kind:       "Deployment",
+			APIVersion: "apps/v1",
+		},
+		ObjectMeta: metav1.ObjectMeta{
+			GenerateName: fmt.Sprintf("provider-%s-", provider.Name),
+			Namespace:    provider.Namespace,
+			Labels:       deploymentLabels,
+		},
+		Spec: appsv1.DeploymentSpec{
+			Selector: &metav1.LabelSelector{
+				MatchLabels: matchLabels,
+			},
+			Replicas: &replicas,
+			Template: podTemplate,
+		},
+	}
+
+	if err := setResourceHashAnnotation(deployment); err != nil {
+		return nil, err
+	}
+
+	return deployment, nil
+}
+
+func populateServiceContainer(serviceContainerName string, podTemplate v1.PodTemplateSpec, provider *pipelinesv1.Provider) v1.PodTemplateSpec {
+	for i, container := range podTemplate.Spec.Containers {
+		if container.Name == serviceContainerName {
+			podTemplate.Spec.Containers[i].Image = provider.Spec.ServiceImage
+			podTemplate.Spec.Containers[i].Env = append(podTemplate.Spec.Containers[i].Env, v1.EnvVar{
+				Name:  "PROVIDERNAME",
+				Value: provider.Name,
+			})
+		}
+	}
+	return podTemplate
+}
+
+func setResourceHashAnnotation(deployment *appsv1.Deployment) error {
+	hasher := pipelines.NewObjectHasher()
+	err := hasher.WriteObject(deployment)
+	if err != nil {
+		return err
+	}
```

Comment:
```suggestion
	if err := hasher.WriteObject(deployment); err != nil {
	    return err
	}
```

Reply:
quite a few places within this file that are doing the same

Reply:
Fixed all in https://github.com/sky-uk/kfp-operator/commit/8b421e577d6992d542f07803bb55b3e9f1cea44a